### Pass / Fail Distribution

- Pass (-1): 1,463 samples (93.36%)
- Fail (1): 104 samples (6.64%)
- Approximate class ratio: 14 : 1

Fail samples account for only about 6.6% of the dataset, confirming a strong class imbalance.
Therefore, model performance should not be evaluated using accuracy alone.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../01_raw/uci-secom.csv")

In [3]:
print("Dataset Shape:", df.shape)

Dataset Shape: (1567, 592)


In [4]:
display(df.head())

,Time,0,1,2,3,4,5,6,7,8,...,581,582,583,584,585,586,587,588,589,Pass/Fail
0,2008-07-19 11:55:00,3030.93,2564.00,2187.7333,1411.1265,1.3602,100.0,97.6133,0.1242,1.5005,...,NaN,0.5005,0.0118,0.0035,2.3630,NaN,NaN,NaN,NaN,-1
1,2008-07-19 12:32:00,3095.78,2465.14,2230.4222,1463.6606,0.8294,100.0,102.3433,0.1247,1.4966,...,208.2045,0.5019,0.0223,0.0055,4.4447,0.0096,0.0201,0.0060,208.2045,-1
2,2008-07-19 13:17:00,2932.61,2559.94,2186.4111,1698.0172,1.5102,100.0,95.4878,0.1241,1.4436,...,82.8602,0.4958,0.0157,0.0039,3.1745,0.0584,0.0484,0.0148,82.8602,1
3,2008-07-19 14:43:00,2988.72,2479.90,2199.0333,909.7926,1.3204,100.0,104.2367,0.1217,1.4882,...,73.8432,0.4990,0.0103,0.0025,2.0544,0.0202,0.0149,0.0044,73.8432,-1
4,2008-07-19 15:22:00,3032.24,2502.87,2233.3667,1326.5200,1.5334,100.0,100.3967,0.1235,1.5031,...,NaN,0.4800,0.4766,0.1045,99.3032,0.0202,0.0149,0.0044,73.8432,-1


print(df.columns.tolist())

In [6]:
# Pass / Fail 개수 확인

In [7]:
print(df["Pass/Fail"].value_counts())

Pass/Fail
-1    1463
 1     104
Name: count, dtype: int64


In [9]:
# Pass / Fail 비울 확인
print(df["Pass/Fail"].value_counts(normalize=True) * 100)

Pass/Fail
-1    93.363114
 1     6.636886
Name: proportion, dtype: float64


In [10]:
# 전체 결측값 개수
total_missing = df.isna().sum().sum()

In [11]:
# 전체 데이터 셀 개수
total_cells = df.shape[0] * df.shape[1]

In [12]:
# 전체 결측률
missing_rate = total_missing / total_cells * 100

In [13]:
print("Total Missing Values:" , total_missing)
print("Total Cells:", total_cells)
print(f"Overall Missing Rate: {missing_rate:.2f}%")

Total Missing Values: 41951
Total Cells: 927664
Overall Missing Rate: 4.52%


In [20]:
# 컬럼별 Missing Rate 계산
missing_rate_by_column = df.isna().mean() * 100

In [21]:
# Missing Rate가 높은 순서대로 정렬
missing_rate_sorted = missing_rate_by_column.sort_values(ascending=False)

display(missing_rate_sorted.head(20))

293    91.193363
292    91.193363
157    91.193363
158    91.193363
492    85.577537
220    85.577537
85     85.577537
358    85.577537
518    64.964901
382    64.964901
245    64.964901
244    64.964901
383    64.964901
384    64.964901
246    64.964901
517    64.964901
110    64.964901
109    64.964901
516    64.964901
111    64.964901
dtype: float64

In [23]:
# 모든 값이 NaN인 컬럼 찾기
all_missing_colums = df.columns[df.isna().all()].tolist()

print("All-Missing Columns:", all_missing_colums)
print("Number of All-Missing Columns:", len(all_missing_colums))

All-Missing Columns: []
Number of All-Missing Columns: 0


In [24]:
# all missing 열이 있는 줄 알았는데 없었음. 한 번 더 검증하기 ( 실제로 엑셀 표를 들어가보니 굉장히 후반부 행에 값이 몇 개 존재했음)
# 결측률이 가장 높은 변수들의 실제 Missing / Nom-Missing 개수 확인
for col in ["293", "292", "157", "158"]:
    print(
        col,
          "| Missing:", df[col].isna().sum(),
        "| Non-Missing:", df[col].notna().sum()
    )

293 | Missing: 1429 | Non-Missing: 138
292 | Missing: 1429 | Non-Missing: 138
157 | Missing: 1429 | Non-Missing: 138
158 | Missing: 1429 | Non-Missing: 138


In [25]:
# 전체 데이터에서 가장 많은 결측 개수가 얼마인지도 확인- 더블체크
print("Maximum Missing Count:", df.isna().sum().max())
print("Total Rows:", len(df))

Maximum Missing Count: 1429
Total Rows: 1567


In [28]:
#센서 컬럼만 선택
sensor_cols = [str(i) for i in range(590)]

# 모든 행의 값이 실제 숫자 0 인 컬럼
all_zero_columns = [
    col for col in sensor_cols
    if df[col].notna().all() and (df[col] == 0).all()
]

print("All-Zero Columns:", all_zero_columns)
print("Number of All-Zero Columns:", len(all_zero_columns))

All-Zero Columns: []
Number of All-Zero Columns: 0


In [29]:
# NaN을 제외했을 때 관측값 종류가 1개인 센서 컬럼 찾기
observed_constant_columns = [
    col for col in sensor_cols
    if df[col].dropna().nunique() == 1
]

print("Observed-value Constant Columns:", observed_constant_columns)
print("Number of Observed-value Constant Columns:", len(observed_constant_columns))

Observed-value Constant Columns: ['5', '13', '42', '49', '52', '69', '97', '141', '149', '178', '179', '186', '189', '190', '191', '192', '193', '194', '226', '229', '230', '231', '232', '233', '234', '235', '236', '237', '240', '241', '242', '243', '256', '257', '258', '259', '260', '261', '262', '263', '264', '265', '266', '276', '284', '313', '314', '315', '322', '325', '326', '327', '328', '329', '330', '364', '369', '370', '371', '372', '373', '374', '375', '378', '379', '380', '381', '394', '395', '396', '397', '398', '399', '400', '401', '402', '403', '404', '414', '422', '449', '450', '451', '458', '461', '462', '463', '464', '465', '466', '481', '498', '501', '502', '503', '504', '505', '506', '507', '508', '509', '512', '513', '514', '515', '528', '529', '530', '531', '532', '533', '534', '535', '536', '537', '538']
Number of Observed-value Constant Columns: 116


In [30]:
# 관측값 1개인게 대체 무슨 값이고, 각 컬럼에는 NaN이 얼마나 섞여 있는가?

constant_summary = []

for col in observed_constant_columns:
    observed_value = df[col].dropna().iloc[0]
    missing_count = df[col].isna().sum()
    missing_rate = df[col].isna().mean() * 100

    constant_summary.append({
        "Feature": col,
        "Observed_Value": observed_value,
        "Missing_Count": missing_count,
        "Missing_Rate(%)": missing_rate
    })

constant_summary_df = pd.DataFrame(constant_summary)

display(constant_summary_df.head(20))

,Feature,Observed_Value,Missing_Count,Missing_Rate(%)
0,5,100.0,14,0.893427
1,13,0.0,3,0.191449
2,42,70.0,1,0.063816
3,49,1.0,1,0.063816
4,52,0.0,1,0.063816
5,69,1.0,6,0.382897
6,97,0.0,6,0.382897
7,141,0.0,14,0.893427
8,149,0.0,3,0.191449
9,178,0.0,24,1.531589


In [31]:
# NaN까지 포함해서 실제로 모든 행이 동일한 Strict Constant Feature
strict_constant_columns = [
    col for col in sensor_cols
    if df[col].notna().all() and df[col].nunique() == 1
]

# 관측값은 하나지만 일부 NaN이 존재하는 Feature
constant_with_missing_columns = [
    col for col in observed_constant_columns
    if df[col].isna().any()
]

print("Strict Constant Columns:", strict_constant_columns)
print("Number of Strict Constant Columns:", len(strict_constant_columns))

print()

print("Observed Constant + Missing Columns:", constant_with_missing_columns)
print("Number of Observed Constant + Missing Columns:", len(constant_with_missing_columns))

Strict Constant Columns: []
Number of Strict Constant Columns: 0

Observed Constant + Missing Columns: ['5', '13', '42', '49', '52', '69', '97', '141', '149', '178', '179', '186', '189', '190', '191', '192', '193', '194', '226', '229', '230', '231', '232', '233', '234', '235', '236', '237', '240', '241', '242', '243', '256', '257', '258', '259', '260', '261', '262', '263', '264', '265', '266', '276', '284', '313', '314', '315', '322', '325', '326', '327', '328', '329', '330', '364', '369', '370', '371', '372', '373', '374', '375', '378', '379', '380', '381', '394', '395', '396', '397', '398', '399', '400', '401', '402', '403', '404', '414', '422', '449', '450', '451', '458', '461', '462', '463', '464', '465', '466', '481', '498', '501', '502', '503', '504', '505', '506', '507', '508', '509', '512', '513', '514', '515', '528', '529', '530', '531', '532', '533', '534', '535', '536', '537', '538']
Number of Observed Constant + Missing Columns: 116


In [33]:
# RAW Sensor Data에서 실제 숫자 0확인

sensor_df = df[sensor_cols]

# 전체 0 개수
total_zero_count = (sensor_df == 0).sum().sum()

# 0이 한 번이라도 증장하는 Feature 개수
features_with_zero = (sensor_df == 0).any().sum()

print("Total Zero Values:", total_zero_count)
print("Features Containing Zero:", features_with_zero)

Total Zero Values: 200005
Features Containing Zero: 215


In [34]:
print("Time Missing:", df["Time"].isna().sum())
print("Pass/Fail Missing:", df["Pass/Fail"].isna().sum())

print()
print("Pass/Fail Unique Values:", sorted(df["Pass/Fail"].dropna().unique()))

Time Missing: 0
Pass/Fail Missing: 0

Pass/Fail Unique Values: [np.int64(-1), np.int64(1)]


## Phase 1. Raw Data Audit Summary

-Dataset: 1,567 samples * 592 columns
-Process features: 590 anonymous measurement variables(secom 데이터의 본래 가진 특성. 변수의 익명화)
-Pass (-1): 1,463 samples (93.36%)
-Fail (1): 104 samples (6.64%)
-Overall missing values: 41,951 (4.52%) 결측치
-Maximum feature-level missing rate: 91.19% (한 열의 91.19%가 결측치임. 모든 행이 NaN인 열은 존재하지 않음)
-All-missing features: 0
-Observed-value constant features with missing values: 116(결측치를 제외한 나머지 행이 모두 같은 값인 열의 개수)
-Raw zero values in process features: 200,005 (21.6%) 전체 데이터셋에서, 값이 0인 셀의 개수
-Features containing at least one zero: 215
-Missing values in Time: 0 (시간 열에는 문제없다)
-Missing values in Pass/ Fail: 0 
-Target values: -1, 1

### Key Findings

1. The dataset shows a strong class imbalance, with Fail samples accounting for approximately 6.64%.
2. Although the overall missing rate is 4.52%, missing values are highly concentrated in some features, with the highest missing rate exceeding 91%.
3. No all-missing or strict constant features were identified.
4. 116 features have only one observed numerical value but also contain missing values. These features will not be automatically removed because their missingness patterns may contain predictive information.
5. Raw process data already contains a substantial number of numerical zero values. Therefore, missing values must be distinguished from actual zero values during preprocessing.